File: go/src/utils/mfile/FileUtils.go
package FileUtils

import (
	"bufio"
	"io"
	"io/ioutil"
	"os"
	"strings"
)

const (
	NEW_FILE_PERM = 0644 // FileMode
)

/**
 * 判断文件是否存在  存在返回 true 不存在返回false
 */
func Exists(filename string) bool {
	var exist = true
	if _, err := os.Stat(filename); os.IsNotExist(err) {
		exist = false
	}
	return exist
}

func IsFile(filename string) bool {
	finfo, err := os.Stat(filename)
	if err != nil {
		return false
	}
	if finfo.IsDir() {
		return false
	} else {
		return true
	}

}
func IsDir(filename string) bool {
	finfo, err := os.Stat(filename)
	if err != nil {
		return false
	}
	if finfo.IsDir() {
		return true
	} else {
		return false
	}

}

// 返回文件和目录，不包含路径，只有名字
func ListDir(pathname string) []string {
	files := []string{}
	dir, err := ioutil.ReadDir(pathname)
	if err != nil {
		return nil
	}
	for _, fi := range dir {
		//if fi.IsDir() { // 忽略目录
		//	continue
		//}
		files = append(files, fi.Name())
	}
	return files
}

// defer 声明的操作在return之前 逆序执行。在go语言里，记得open一个文件做了错误判断后要紧跟一个defer close延迟调用

// 复制文件，dest如果存在会被覆盖。
func CopyFile(src, dest string) error {
	sp, err := os.Open(src)
	if err != nil {
		return err
	}
	defer sp.Close()

	dp, err := os.OpenFile(dest, os.O_WRONLY|os.O_CREATE, 0644)
	if err != nil {
		return err
	}
	defer dp.Close()

	_, err = io.Copy(dp, sp)
	if err != nil {
		return err
	}
	return nil
}

// 移动文件
func MoveFile(src, dest string) error {
	return os.Rename(src, dest)
}

// 删除文件、目录（要求目录内不能有文件存在）
func Remove(filename string) error {
	return os.Remove(filename)
}

// 删除文件、文件夹，强制删除
func RemoveAll(path string) error {
	return os.RemoveAll(path)
}

// 创建文件夹，不允许级联创建。 path原文件夹存在不会报错，
func Mkdir(path string) error {
	return os.Mkdir(path, os.ModePerm)
}

// 创建文件夹，允许级联创建。  path原文件夹存在不会报错，
func MkdirAll(path string) error {
	return os.MkdirAll(path, os.ModePerm)
}

// Read reads the file named by filename and returns the contents.
func Read(filename string) ([]byte, error) {
	return ioutil.ReadFile(filename)
}

// ReadString reads the file named by filename and returns the contents as string.
func ReadString(filename string) (string, error) {
	buf, err := Read(filename)
	if err != nil {
		return "", err
	} else {
		return string(buf), nil
	}
}

func ReadLines(filename string) ([]string, error) {
	f, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var cont []string
	rd := bufio.NewReader(f)
	for {
		line, err := rd.ReadString('\n') //以'\n'为结束符读入一行

		if err != nil || io.EOF == err {
			break
		}
		cont = append(cont, strings.TrimRight(line, "\r\n"))
	}
	return cont, nil
}

// Write writes data to a file named by filename.
func Write(filename string, content []byte) error {
	return ioutil.WriteFile(filename, content, NEW_FILE_PERM)
}

// 会覆盖原文件
func WriteString(filename, content string) error {
	return writeString(filename, content, false)
}

// 向文件中追加内容
func AppendString(filename, content string) error {
	return writeString(filename, content, true)
}

func writeString(filename, content string, append bool) error {
	var flag int
	if append {
		flag = os.O_WRONLY | os.O_CREATE | os.O_APPEND
	} else {
		flag = os.O_WRONLY | os.O_CREATE | os.O_TRUNC
	}
	f, err := os.OpenFile(filename, flag, NEW_FILE_PERM)
	if err != nil {
		return err
	}
	data := []byte(content)
	n, err := f.Write(data)
	if err == nil && n < len(data) {
		err = io.ErrShortWrite
	}
	if err1 := f.Close(); err == nil {
		err = err1
	}
	return err
}

// 同linux touch命令，如果原文件存在则不处理； 要求父目录必须存在
func Touch(filename string) error {
	_, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_APPEND, NEW_FILE_PERM)
	if err != nil {
		return err
	}
	return err
}

// 会覆盖原文件
func WriteLines(filename string, contents []string) error {
	return writeLines(filename, contents, false)
}

// 向文件中追加内容
func AppendLines(filename string, contents []string) error {
	return writeLines(filename, contents, true)
}

func writeLines(filename string, contents []string, append bool) error {
	var flag int
	if append {
		flag = os.O_WRONLY | os.O_CREATE | os.O_APPEND
	} else {
		flag = os.O_WRONLY | os.O_CREATE | os.O_TRUNC
	}
	f, err := os.OpenFile(filename, flag, NEW_FILE_PERM)
	if err != nil {
		return err
	}
	for _, content := range contents {
		data := []byte(content)
		f.Write(data)
		f.Write([]byte("\n"))
	}

	if err1 := f.Close(); err == nil {
		err = err1
	}
	return err
}

File: python/utils/mfile/FileUtils.py
# -*- coding: utf-8 -*-
import shutil
import os
import os.path

'''
os.path.dirname(path)  返回path的父目录

# 遍历文件夹下文件
rootdir = "/Users/kongzhihui/test"  # 指明被遍历的文件夹

    for parent, dirnames, filenames in os.walk(rootdir):  # 三个参数：分别返回1.父目录 2.所有文件夹名字（不含路径） 3.所有文件名字
        for dirname in dirnames:  # 输出文件夹信息
            print "parent is:" + parent
            print  "dirname is:" + dirname

        for filename in filenames:  # 输出文件信息
            print "parent is:"+ parent
            print "filename is:" + filename
            print "the full name of the file is:" + os.path.join(parent, filename)  # 输出文件路径信息
'''


def exists(path):
    return os.path.exists(path)


def isFile(path):
    return os.path.isfile(path)


def isDir(path):
    return os.path.isdir(path)


# 返回文件和目录，不包含路径，只有名字
def listdir(path):
    return os.listdir(path)


# 要求父目录必须存在
def touch(fname):
    fp = open(fname, 'a')
    fp.close()


# 要求src必须存在
def copyfile(src, dest):
    shutil.copy(src, dest)


# 要求src必须存在
def moveFile(src, dest):
    shutil.move(src, dest)


# 要求fname必须存在，fname必须是文件
def deleteFile(fname):
    os.remove(fname)


# 要求src必须存在，dest必须不存在； 目录下文件也会复制
def copyDirectory(src, dest):
    shutil.copytree(src, dest)


# 要求dname不存在，force=True会自动创建起父目录
def mkdir(dname, force=False):
    if force:
        os.makedirs(dname)
    else:
        os.mkdir(dname)


# 要求dname必须存在，force=False时候要求dname为空目录
def rmdir(dname, force=False):
    if force:
        shutil.rmtree(dname)
    else:
        os.rmdir(dname)


# 返回str，要求fname必须存在，utf-8编码
# 读取非UTF-8编码的文本文件，需要给open()函数传入encoding参数，例如，读取GBK编码的文件：f = open('/Users/michael/gbk.txt', 'r', encoding='gbk')
# 遇到有些编码不规范的文件，可能会遇到UnicodeDecodeError。遇到这种情况，open()函数还接收一个errors参数：f = open('/Users/michael/gbk.txt', 'r', encoding='gbk', errors='ignore')

# 对于二进制文件，如图片，需要制定使用 字节形式来读取，open(fname, "rb")，否则读取信息会丢失。

def readFileToString(fname):
    with open(fname) as fp:  # 或者fp = open("file.txt","r")
        cont = fp.read()
    # fp = None
    # cont = None
    # try:
    #     fp = open(fname)  # 或者fp = open("file.txt","r")
    #     cont = fp.read()
    # finally:
    #     if fp:
    #         fp.close()

    return cont


'''
for line in open(fname):
   print line.strip()
'''


# 返回list，每行删除回车，要求fname必须存在；  fp.readlines() 也返回list，但是每行会带上回车
def readLines(fname):
    ret = []
    with open(fname) as fp:
        for line in fp:
            ret.append(line.rstrip("\r\n"))
    return ret


# 会把原来的文件内容
def writeStringToFile(fname, s, append=False):
    mode = "w"
    if append:
        mode = "a"

    with open(fname, mode) as fp:
        fp.write(s)


# l为list，每个元素不用有回车，输出到文件的时候每个item会自动补上回车。
def writeLines(fname, l, append=False):
    mode = "w"
    if append:
        mode = "a"
    with open(fname, mode) as fp:
        for i in l:
            fp.write(str(i))
            fp.write("\n")
